File: apps/backend/src/rhesis/backend/app/schemas/multi_turn_test_config.py
"""
Schema for multi-turn test configuration stored in test.test_configuration JSON field.

This is separate from the TestConfiguration entity - this validates the JSON
stored in the test's test_configuration attribute for multi-turn tests.
"""

from typing import Optional

from pydantic import BaseModel, Field, field_validator


class MultiTurnTestConfig(BaseModel):
    """
    Configuration for multi-turn tests.

    Stored in test.test_configuration JSONB column when test_type is multi-turn.
    These fields define how Penelope should execute the test.
    """

    goal: str = Field(
        ...,
        min_length=1,
        max_length=5000,
        description="What the target SHOULD do - success criteria (required)",
    )
    instructions: Optional[str] = Field(
        None,
        max_length=10000,
        description="HOW Penelope should conduct the test - methodology (optional)",
    )
    restrictions: Optional[str] = Field(
        None,
        max_length=10000,
        description="What the target MUST NOT do - forbidden behaviors (optional)",
    )
    scenario: Optional[str] = Field(
        None,
        max_length=5000,
        description="Context and persona for the test (optional)",
    )
    max_turns: Optional[int] = Field(
        default=10,
        ge=1,
        le=50,
        description="Maximum number of conversation turns (default: 10)",
    )

    @field_validator("goal", "instructions", "restrictions", "scenario")
    @classmethod
    def strip_whitespace(cls, v: Optional[str]) -> Optional[str]:
        """Strip leading/trailing whitespace from string fields."""
        if v is None:
            return None
        stripped = v.strip()
        return stripped if stripped else None

    model_config = {"extra": "forbid"}  # Don't allow additional fields


def validate_multi_turn_config(config_dict: dict) -> MultiTurnTestConfig:
    """
    Validate a test_configuration dict for multi-turn tests.

    Args:
        config_dict: Dictionary to validate

    Returns:
        Validated MultiTurnTestConfig instance

    Raises:
        ValidationError: If validation fails
    """
    return MultiTurnTestConfig.model_validate(config_dict)

File: apps/frontend/src/utils/api-client/interfaces/multi-turn-test-config.ts
/**
 * Multi-turn test configuration interfaces.
 *
 * These define the structure stored in test.test_configuration JSON
 * for multi-turn tests (when test_type is multi-turn).
 */

export interface MultiTurnTestConfig {
  /**
   * What the target SHOULD do - success criteria (required)
   */
  goal: string;

  /**
   * HOW Penelope should conduct the test - methodology (optional)
   */
  instructions?: string;

  /**
   * What the target MUST NOT do - forbidden behaviors (optional)
   */
  restrictions?: string;

  /**
   * Context and persona for the test (optional)
   */
  scenario?: string;

  /**
   * Maximum number of conversation turns (default: 10, range: 1-50)
   */
  max_turns?: number;
}

/**
 * Type guard to check if a test_configuration is a multi-turn config
 * Validates both presence and types of fields
 */
export function isMultiTurnConfig(config: any): config is MultiTurnTestConfig {
  if (config === null || typeof config !== 'object' || !('goal' in config)) {
    return false;
  }

  // Validate required field type
  if (typeof config.goal !== 'string') {
    return false;
  }

  // Validate optional field types if present
  if ('instructions' in config && typeof config.instructions !== 'string') {
    return false;
  }

  if ('restrictions' in config && typeof config.restrictions !== 'string') {
    return false;
  }

  if ('scenario' in config && typeof config.scenario !== 'string') {
    return false;
  }

  if ('max_turns' in config && typeof config.max_turns !== 'number') {
    return false;
  }

  return true;
}

/**
 * Create an empty multi-turn test configuration with defaults
 */
export function createEmptyMultiTurnConfig(): MultiTurnTestConfig {
  return {
    goal: '',
    instructions: undefined,
    restrictions: undefined,
    scenario: undefined,
    max_turns: 10,
  };
}
